blmath/geometry/primitives/plane.py

```diff
@@ -221,157 +260,111 @@ def mesh_xsection(self, m, neighborhood=None):
-        #   [0, 1], [2, 3], [4, 5], ...
-        # But should contain
-        #   [0, 1], [1, 2], [2, 3], ...
-        # Leaving in place since the code below may depend on it.
-        e = np.array([[i, i + 1] for i in xrange(0, v.shape[0], 2)])
-
-        # Step 4 (optional - only if 'neighborhood' is provided):
-        #   Build and return the ordered vertices cmp_v, and the
-        #   edges cmp_e. Get connected components, use a KDTree
-        #   to select the one with minimal distance to 'component'.
-        #   Return the cmp_v and (re-indexed) edge mapping cmp_e.
+        if len(fs) == 0:
+            return [] # Nothing intersects
+
+        # 2: Find the edges where each of those faces actually cross the plane
+        def edge_from_face(f):
+            face_verts = [
+                [m.v[f[0]], m.v[f[1]]],
+                [m.v[f[1]], m.v[f[2]]],
+                [m.v[f[2]], m.v[f[0]]],
+            ]
+            e = [self.line_segment_xsection(a, b) for a, b in face_verts]
+            e = [x for x in e if x is not None]
+            return e
+        edges = np.vstack([np.hstack(edge_from_face(f)) for f in fs])
+
+        # 3: Find the set of unique vertices in `edges`
+        v1s, v2s = np.hsplit(edges, 2)
+        verts = edges.reshape((-1, 3))
+        verts = np.vstack(sorted(verts, key=operator.itemgetter(0, 1, 2)))
+        eps = 1e-15 # the floating point calculation of the intersection locations is not _quite_ exact
+        verts = verts[list(np.sqrt(np.sum(np.diff(verts, axis=0) ** 2, axis=1)) > eps) + [True]]
+        # the True at the end there is because np.diff returns pairwise differences; one less element than the original array
+
+        # 4: Build the edge adjacency matrix
+        E = sp.dok_matrix((verts.shape[0], verts.shape[0]), dtype=np.bool)
+        def indexof(v, in_this):
+            return np.nonzero(np.all(np.abs(in_this - v) < eps, axis=1))[0]
+        for ii, v in enumerate(verts):
+            for other_v in list(v2s[indexof(v, v1s)]) + list(v1s[indexof(v, v2s)]):
+                neighbors = indexof(other_v, verts)
+                E[ii, neighbors] = True
+                E[neighbors, ii] = True
+
+        def eulerPath(E):
+            # Based on code from Przemek Drochomirecki, Krakow, 5 Nov 2006
+            # http://code.activestate.com/recipes/498243-finding-eulerian-path-in-undirected-graph/
+            # Under PSF License
+            # NB: MUTATES graph
+            if len(E.nonzero()[0]) == 0:
+                return None
+            # counting the number of vertices with odd degree
+            odd = list(np.nonzero(np.bitwise_and(np.sum(E, axis=0), 1))[1])
+            odd.append(np.nonzero(E)[0][0])
+            # This check is appropriate if there is a single connected component.
+            # Since we're willing to take away one connected component per call,
+            # we skip this check.
+            # if len(odd) > 3:
+            #     return None
```

Comment:
It's relevant to the underlying algorithm; I think leaving it in there makes sense, explaining why we diverge from the algorithm named.